src/catalog/pgduckdb_table.cpp

```diff
@@ -25,18 +27,29 @@ extern "C" {
 
 namespace duckdb {
 
-PostgresTable::PostgresTable(Catalog &catalog, SchemaCatalogEntry &schema, CreateTableInfo &info,
+PostgresTable::PostgresTable(Catalog &catalog, SchemaCatalogEntry &schema, CreateTableInfo &info, ::Relation rel,
                              Cardinality cardinality, Snapshot snapshot)
-    : TableCatalogEntry(catalog, schema, info), cardinality(cardinality), snapshot(snapshot) {
+    : TableCatalogEntry(catalog, schema, info), rel(rel), cardinality(cardinality), snapshot(snapshot) {
+}
+
+PostgresTable::~PostgresTable() {
+	std::lock_guard<std::mutex> lock(pgduckdb::DuckdbProcessLock::GetLock());
+	RelationClose(rel);
+}
+
+::Relation
+PostgresTable::OpenRelation(Oid relid) {
+	std::lock_guard<std::mutex> lock(pgduckdb::DuckdbProcessLock::GetLock());
+	auto rel = pgduckdb::PostgresFunctionGuard<::Relation>(RelationIdGetRelation, relid);
+	return rel;
 }
 
 bool
-PostgresTable::SetTableInfo(CreateTableInfo &info, Oid relid, Snapshot snapshot) {
-	auto rel = RelationIdGetRelation(relid);
+PostgresTable::SetTableInfo(CreateTableInfo &info, ::Relation rel) {
 	auto tupleDesc = RelationGetDescr(rel);
 
 	if (!tupleDesc) {
-		elog(WARNING, "Failed to get tuple descriptor for relation with OID %u", relid);
+		elog(WARNING, "Failed to get tuple descriptor for relation with OID %u", rel->rd_id);
 		RelationClose(rel);
```

Comment:
This one should now be removed I think.

Reply:
Yes, thanks. I'm thinking if this check is even needed? If we have relation object opened before should we always have tuple desc information?

Reply:
Yeah, that sounds reasonable. There's no way for this to fail. It's a simple macro that reads a field:
```
#define RelationGetDescr(relation) ((relation)->rd_att)
```